File: backend/app/services/progression.py
"""Adaptive progression: level unlocks, concept mastery, and remediation.

Rules:
- Passing ``PASSES_TO_UNLOCK`` distinct exercises at the highest unlocked level
  unlocks the next level.
- Failing (or giving up on) the same exercise ``CONSECUTIVE_FAILS_FOR_REMEDIATION``
  times in a row opens a remediation: the trainer suggests other exercises that share
  the concept at the same or lower level until ``REMEDIATION_PASSES_REQUIRED`` of them
  are passed, then steers the student back to the exercise they were stuck on.
- A concept is considered mastered after ``MASTERY_PASSES`` passed exercises.
"""

from dataclasses import dataclass, field

from sqlalchemy import func, select
from sqlalchemy.orm import Session

from ..models import Attempt, ConceptMastery, Exercise, Remediation, User, UserProgress

PASSES_TO_UNLOCK = 2
CONSECUTIVE_FAILS_FOR_REMEDIATION = 3
REMEDIATION_PASSES_REQUIRED = 2
MASTERY_PASSES = 2


@dataclass
class OutcomeSummary:
    newly_unlocked_level: int | None = None
    remediation_started: bool = False
    remediation_active: bool = False
    remediation_completed: bool = False
    messages: list[str] = field(default_factory=list)

    @property
    def message(self) -> str:
        return " ".join(self.messages)


def get_or_create_progress(db: Session, user: User) -> UserProgress:
    progress = db.get(UserProgress, user.id)
    if progress is None:
        progress = UserProgress(user_id=user.id, unlocked_level=1)
        db.add(progress)
        db.flush()
    return progress


def get_or_create_mastery(db: Session, user_id: int, concept: str) -> ConceptMastery:
    mastery = db.execute(
        select(ConceptMastery).where(
            ConceptMastery.user_id == user_id, ConceptMastery.concept == concept
        )
    ).scalar_one_or_none()
    if mastery is None:
        mastery = ConceptMastery(user_id=user_id, concept=concept, passes=0, fails=0)
        db.add(mastery)
        db.flush()
    return mastery


def max_level(db: Session) -> int:
    result = db.execute(
        select(func.max(Exercise.level)).where(Exercise.is_active.is_(True))
    ).scalar()
    return result or 1


def passed_exercise_ids(db: Session, user_id: int) -> set[int]:
    rows = db.execute(
        select(Attempt.exercise_id)
        .where(Attempt.user_id == user_id, Attempt.passed.is_(True))
        .distinct()
    ).all()
    return {row[0] for row in rows}


def consecutive_failures(db: Session, user_id: int, exercise_id: int) -> int:
    attempts = db.execute(
        select(Attempt.passed)
        .where(Attempt.user_id == user_id, Attempt.exercise_id == exercise_id)
        .order_by(Attempt.id.desc())
    ).all()
    count = 0
    for (passed,) in attempts:
        if passed:
            break
        count += 1
    return count


def active_remediation(db: Session, user_id: int) -> Remediation | None:
    return db.execute(
        select(Remediation)
        .where(Remediation.user_id == user_id, Remediation.active.is_(True))
        .order_by(Remediation.id.desc())
    ).scalars().first()


def latest_completed_remediation(db: Session, user_id: int) -> Remediation | None:
    return db.execute(
        select(Remediation)
        .where(
            Remediation.user_id == user_id,
            Remediation.active.is_(False),
            Remediation.completed_passes >= REMEDIATION_PASSES_REQUIRED,
        )
        .order_by(Remediation.id.desc())
    ).scalars().first()


def record_attempt_outcome(
    db: Session, user: User, exercise: Exercise, passed: bool
) -> OutcomeSummary:
    """Update mastery, remediation, and level unlocks after an attempt is stored."""
    summary = OutcomeSummary()
    progress = get_or_create_progress(db, user)
    mastery = get_or_create_mastery(db, user.id, exercise.concept)
    remediation = active_remediation(db, user.id)

    if passed:
        mastery.passes += 1
        if remediation is not None and remediation.concept == exercise.concept and (
            remediation.source_exercise_id != exercise.id
        ):
            remediation.completed_passes += 1
            if remediation.completed_passes >= REMEDIATION_PASSES_REQUIRED:
                remediation.active = False
                summary.remediation_completed = True
                summary.messages.append(
                    "Great work — you've practiced this concept enough. "
                    "Time to go back to the exercise you were stuck on."
                )
        if remediation is not None and remediation.source_exercise_id == exercise.id:
            remediation.active = False
            summary.remediation_completed = True
            summary.messages.append("You beat the exercise you were stuck on!")

        newly_unlocked = _check_level_unlock(db, user, progress, exercise)
        if newly_unlocked is not None:
            summary.newly_unlocked_level = newly_unlocked
            summary.messages.append(f"Level {newly_unlocked} unlocked!")
    else:
        mastery.fails += 1
        fails = consecutive_failures(db, user.id, exercise.id)
        if fails >= CONSECUTIVE_FAILS_FOR_REMEDIATION and remediation is None:
            candidates = remediation_candidates(db, user.id, exercise)
            if candidates:
                db.add(
                    Remediation(
                        user_id=user.id,
                        source_exercise_id=exercise.id,
                        concept=exercise.concept,
                        completed_passes=0,
                        active=True,
                    )
                )
                db.flush()
                summary.remediation_started = True
                summary.messages.append(
                    "This one is giving you trouble — let's practice some similar "
                    "rhythms first and come back to it."
                )

    summary.remediation_active = active_remediation(db, user.id) is not None
    return summary


def _check_level_unlock(
    db: Session, user: User, progress: UserProgress, exercise: Exercise
) -> int | None:
    if exercise.level != progress.unlocked_level:
        return None
    highest = max_level(db)
    if progress.unlocked_level >= highest:
        return None
    passed_at_level = db.execute(
        select(func.count(func.distinct(Attempt.exercise_id)))
        .select_from(Attempt)
        .join(Exercise, Exercise.id == Attempt.exercise_id)
        .where(
            Attempt.user_id == user.id,
            Attempt.passed.is_(True),
            Exercise.level == progress.unlocked_level,
        )
    ).scalar()
    if (passed_at_level or 0) >= PASSES_TO_UNLOCK:
        progress.unlocked_level += 1
        return progress.unlocked_level
    return None


def remediation_candidates(
    db: Session, user_id: int, source: Exercise
) -> list[Exercise]:
    """Unpassed, active exercises sharing the source's concept at the same or lower level."""
    passed = passed_exercise_ids(db, user_id)
    rows = db.execute(
        select(Exercise)
        .where(
            Exercise.is_active.is_(True),
            Exercise.concept == source.concept,
            Exercise.level <= source.level,
            Exercise.id != source.id,
        )
        .order_by(Exercise.level.desc(), Exercise.id.asc())
    ).scalars().all()
    return [e for e in rows if e.id not in passed]


def next_exercise(db: Session, user: User) -> tuple[Exercise | None, str, str]:
    """Return ``(exercise, reason, message)`` for the student's recommended next step."""
    progress = get_or_create_progress(db, user)
    passed = passed_exercise_ids(db, user.id)

    remediation = active_remediation(db, user.id)
    if remediation is not None:
        source = db.get(Exercise, remediation.source_exercise_id)
        if source is not None:
            candidates = remediation_candidates(db, user.id, source)
            if candidates:
                remaining = REMEDIATION_PASSES_REQUIRED - remediation.completed_passes
                return (
                    candidates[0],
                    "remediation",
                    f"Let's practice a similar rhythm. Pass {remaining} more like this "
                    f"and we'll go back to “{source.title}”.",
                )
        # Nothing left to practice with — close the remediation and fall through.
        remediation.active = False
        db.flush()

    completed = latest_completed_remediation(db, user.id)
    if completed is not None:
        source = db.get(Exercise, completed.source_exercise_id)
        if source is not None and source.is_active and source.id not in passed:
            return (
                source,
                "progression",
                "You've warmed up — time to retry the exercise you were stuck on.",
            )

    candidates = db.execute(
        select(Exercise)
        .where(Exercise.is_active.is_(True), Exercise.level <= progress.unlocked_level)
        .order_by(Exercise.level.asc(), Exercise.id.asc())
    ).scalars().all()
    for exercise in candidates:
        if exercise.id not in passed:
            return exercise, "progression", "Here's your next exercise."

    return None, "complete", "You've passed every unlocked exercise. Nice work!"

File: frontend/src/lib/progression.ts
/**
 * TypeScript port of backend/app/services/progression.py
 * Adaptive level unlocks, concept mastery, and remediation — all in localStorage.
 */

import type { Exercise } from '../api/types'
import {
  addAttempt,
  getAttempts,
  getMastery,
  getProgress,
  getRemediation,
  nextExerciseId,
  saveCustomExercises,
  getCustomExercises,
  saveMastery,
  saveProgress,
  saveRemediation,
} from './localDb'
import { tapCount } from './rhythm'
import { SEED_EXERCISES } from '../data/exercises'

const PASSES_TO_UNLOCK = 2
const CONSECUTIVE_FAILS_FOR_REMEDIATION = 3
const REMEDIATION_PASSES_REQUIRED = 2

export interface OutcomeSummary {
  newlyUnlockedLevel: number | null
  remediationStarted: boolean
  remediationActive: boolean
  remediationCompleted: boolean
  message: string
}

export function allExercises(): Exercise[] {
  const custom = getCustomExercises().map((c) => c as unknown as Exercise)
  return [...SEED_EXERCISES, ...custom].filter((e) => e.is_active)
}

export function maxLevel(): number {
  return Math.max(...allExercises().map((e) => e.level), 1)
}

/**
 * Advance unlockedLevel past any levels where all exercises are already passed.
 *
 * This handles the case where new exercises are added to a level above the
 * user's previous maximum: they completed the old max level when it was the
 * ceiling, so checkLevelUnlock never ticked them forward. Calling this at the
 * start of nextExercise() and listExercises() keeps the state correct.
 */
export function catchUpLevel(userId: number): void {
  const progress = getProgress(userId)
  const passed = passedExerciseIds(userId)
  const highest = maxLevel()
  let changed = false

  while (progress.unlockedLevel < highest) {
    const atLevel = allExercises().filter((e) => e.level === progress.unlockedLevel)
    // If every exercise at this level is passed (or there are none), advance.
    if (atLevel.length === 0 || atLevel.every((e) => passed.has(e.id))) {
      progress.unlockedLevel++
      changed = true
    } else {
      break
    }
  }

  if (changed) saveProgress(userId, progress)
}

function passedExerciseIds(userId: number): Set<number> {
  return new Set(
    getAttempts(userId)
      .filter((a) => a.passed)
      .map((a) => a.exerciseId),
  )
}

function consecutiveFailures(userId: number, exerciseId: number): number {
  const attempts = getAttempts(userId)
    .filter((a) => a.exerciseId === exerciseId)
    .reverse()
  let count = 0
  for (const a of attempts) {
    if (a.passed) break
    count++
  }
  return count
}

function remediationCandidates(userId: number, source: Exercise): Exercise[] {
  const passed = passedExerciseIds(userId)
  return allExercises()
    .filter(
      (e) =>
        e.concept === source.concept &&
        e.level <= source.level &&
        e.id !== source.id &&
        !passed.has(e.id),
    )
    .sort((a, b) => b.level - a.level || a.id - b.id)
}

export function recordAttempt(
  userId: number,
  exercise: Exercise,
  tapsMs: number[],
  gaveUp: boolean,
  mode: 'free' | 'strict',
  accuracy: number,
  passed: boolean,
): { attemptId: number; outcome: OutcomeSummary } {
  const attempt = addAttempt(userId, { exerciseId: exercise.id, tapsMs, passed, gaveUp, mode, accuracy })

  const summary: OutcomeSummary = {
    newlyUnlockedLevel: null,
    remediationStarted: false,
    remediationActive: false,
    remediationCompleted: false,
    message: '',
  }

  const mastery = getMastery(userId)
  const cm = mastery[exercise.concept] ?? { passes: 0, fails: 0 }
  const remediation = getRemediation(userId)
  const messages: string[] = []

  if (passed) {
    cm.passes++
    if (
      remediation?.active &&
      remediation.concept === exercise.concept &&
      remediation.sourceExerciseId !== exercise.id
    ) {
      remediation.completedPasses++
      if (remediation.completedPasses >= REMEDIATION_PASSES_REQUIRED) {
        remediation.active = false
        summary.remediationCompleted = true
        messages.push(
          "Great work — you've practiced this concept enough. Time to go back to the exercise you were stuck on.",
        )
        saveRemediation(userId, remediation)
      } else {
        saveRemediation(userId, remediation)
      }
    }
    if (remediation?.active && remediation.sourceExerciseId === exercise.id) {
      remediation.active = false
      summary.remediationCompleted = true
      messages.push('You beat the exercise you were stuck on!')
      saveRemediation(userId, remediation)
    }

    const unlocked = checkLevelUnlock(userId, exercise)
    if (unlocked !== null) {
      summary.newlyUnlockedLevel = unlocked
      messages.push(`Level ${unlocked} unlocked!`)
    }
  } else {
    cm.fails++
    const fails = consecutiveFailures(userId, exercise.id)
    if (fails >= CONSECUTIVE_FAILS_FOR_REMEDIATION && !remediation?.active) {
      const candidates = remediationCandidates(userId, exercise)
      if (candidates.length > 0) {
        saveRemediation(userId, {
          id: Date.now(),
          sourceExerciseId: exercise.id,
          concept: exercise.concept,
          completedPasses: 0,
          active: true,
        })
        summary.remediationStarted = true
        messages.push(
          "This one is giving you trouble — let's practice some similar rhythms first and come back to it.",
        )
      }
    }
  }

  mastery[exercise.concept] = cm
  saveMastery(userId, mastery)

  const currentRemediation = getRemediation(userId)
  summary.remediationActive = currentRemediation?.active ?? false
  summary.message = messages.join(' ')

  return { attemptId: attempt.id, outcome: summary }
}

function checkLevelUnlock(userId: number, exercise: Exercise): number | null {
  const progress = getProgress(userId)
  if (exercise.level !== progress.unlockedLevel) return null
  const highest = maxLevel()
  if (progress.unlockedLevel >= highest) return null

  const attempts = getAttempts(userId)
  const passedAtLevel = new Set(
    attempts
      .filter((a) => a.passed)
      .map((a) => a.exerciseId)
      .filter((eid) => {
        const ex = allExercises().find((e) => e.id === eid)
        return ex?.level === progress.unlockedLevel
      }),
  ).size

  if (passedAtLevel >= PASSES_TO_UNLOCK) {
    progress.unlockedLevel++
    saveProgress(userId, progress)
    return progress.unlockedLevel
  }
  return null
}

export function nextExercise(
  userId: number,
): { exercise: Exercise | null; reason: 'progression' | 'remediation' | 'complete'; message: string } {
  catchUpLevel(userId)
  const progress = getProgress(userId)
  const passed = passedExerciseIds(userId)
  const remediation = getRemediation(userId)

  if (remediation?.active) {
    const source = allExercises().find((e) => e.id === remediation.sourceExerciseId)
    if (source) {
      const candidates = remediationCandidates(userId, source)
      if (candidates.length > 0) {
        const remaining = REMEDIATION_PASSES_REQUIRED - remediation.completedPasses
        return {
          exercise: candidates[0],
          reason: 'remediation',
          message: `Let's practice a similar rhythm. Pass ${remaining} more like this and we'll go back to "${source.title}".`,
        }
      }
    }
    // No candidates left — close remediation and fall through
    if (remediation) {
      remediation.active = false
      saveRemediation(userId, remediation)
    }
  }

  // Check if a recently completed remediation should send us back to the source
  const allRem = getRemediation(userId)
  if (allRem && !allRem.active && allRem.completedPasses >= REMEDIATION_PASSES_REQUIRED) {
    const source = allExercises().find((e) => e.id === allRem.sourceExerciseId)
    if (source && !passed.has(source.id)) {
      return {
        exercise: source,
        reason: 'progression',
        message: 'You\'ve warmed up — time to retry the exercise you were stuck on.',
      }
    }
  }

  const candidates = allExercises()
    .filter((e) => e.level <= progress.unlockedLevel)
    .sort((a, b) => a.level - b.level || a.id - b.id)

  for (const exercise of candidates) {
    if (!passed.has(exercise.id)) {
      return { exercise, reason: 'progression', message: "Here's your next exercise." }
    }
  }

  return { exercise: null, reason: 'complete', message: "You've passed every unlocked exercise. Nice work!" }
}

// Admin: create a custom exercise
export function createCustomExercise(payload: Omit<Exercise, 'id' | 'tap_count'>): Exercise {
  const id = nextExerciseId() + 1000 // offset to avoid collisions with seed IDs
  const exercise: Exercise = { ...payload, id, tap_count: tapCount(payload.pattern) }
  const customs = getCustomExercises()
  customs.push(exercise as unknown as ReturnType<typeof getCustomExercises>[0])
  saveCustomExercises(customs)
  return exercise
}

export function updateCustomExercise(id: number, payload: Omit<Exercise, 'id' | 'tap_count'>): Exercise | null {
  const customs = getCustomExercises()
  const idx = customs.findIndex((e) => e.id === id)
  if (idx === -1) return null
  const updated = { ...payload, id, tap_count: tapCount(payload.pattern) } as unknown as typeof customs[0]
  customs[idx] = updated
  saveCustomExercises(customs)
  return updated as unknown as Exercise
}

export function deleteCustomExercise(id: number): void {
  const customs = getCustomExercises().filter((e) => e.id !== id)
  saveCustomExercises(customs)
}

export function skipToLevel(userId: number, targetLevel: number): void {
  const progress = getProgress(userId)
  if (targetLevel > progress.unlockedLevel) {
    progress.unlockedLevel = targetLevel
    saveProgress(userId, progress)
  }
}
